### ISSUE TITLE
about Discriminator

### ISSUE THREAD

**Issue by budui** [charliememory/Pose-Guided-Person-Image-Generation#28]
Hi, I am porting PG2 to pytorch.
I am confused about the design of Discriminator:
In your code, the last layer of Discriminator is a linear whose output maybe not in [0, 1].
https://github.com/charliememory/Pose-Guided-Person-Image-Generation/blob/fef3e7a3b313c23501e5c2c8178f9ea6bac8ea41/wgan_gp.py#L434

so, I add a sigmoid to output:
```
        self.fc = nn.Linear(8*4*8*base_channels, 1)
        self.sigmoid = nn.Sigmoid()

        self.main.apply(weights_init)

    def forward(self, x):
        x = self.main(x)
        x = x.view(-1, 8*4*8*self.base_channels)
        return self.sigmoid(self.fc(x))
```
copy from [here](https://github.com/budui/Human-Pose-Transfer-For-ReID/blob/6e4a7725b1173abddfb2d75f81bd21ebb1eef1f9/models/PG2.py#L240-L248)

This leads to poor output
![image](https://user-images.githubusercontent.com/16448529/57582259-1f277580-74f5-11e9-8ae8-ed00f0ea261d.png)

if I only change the input layer's stride to (4, 2) of DCGAN's Discriminator.
```
            nn.Conv2d(in_channels=nc, out_channels=nf, kernel_size=4, stride=(4, 2), padding=1, bias=False),
            nn.LeakyReLU(0.2, inplace=True),
```

I get a better output.

![image](https://user-images.githubusercontent.com/16448529/57582282-467e4280-74f5-11e9-8cc9-7ea623c18217.png)

but it's still far from your result in paper.

**Comment by budui** [comment]
My implement  using PyTorch1.0 is here:
https://github.com/budui/Human-Pose-Transfer-For-ReID/blob/master/models/PG2.py

Anyone interested in using pytorch to implement PG2? I need your help!

**Comment by charliememory** [comment]
Hi @budui, many thanks to your effort in the PyTorch version!
The discriminator architecture is borrowed from the improved_wgan_training at that moment as follows
https://github.com/igul222/improved_wgan_training/blob/fa66c574a54c4916d27c55441d33753dcc78f6bc/gan_64x64.py#L428 
Currently, there are new discriminator architectures like PatchGAN.

As to the artifacts, I think it may be caused by the pixel value out of range. Adding a value clamp operation in the output to keep the pixel value between [-1,1] may help to address.
Otherwise, you may need to trade off the loss weights between L1 loss and adversarial loss. And adding perceptual loss will further improve the results.

**Comment by budui** [comment]
Thanks for your reply!
I have tried torch.clamp yesterday, artifacts disappeared, but result is still not good.

I am wondering that how to make the last linear's ouput is in [0,1]?  

Can I email you in Chinese to describe the problem I have encountered?

**Comment by charliememory** [comment]
Sure, my email is liqian.ma@esat.kuleuven.be
